File: js/Map/LinkedListMap.js
class Node {
  constructor(key, value) {
    this.key = key;
    this.value = key;
    this.next = null;
  }

  toString() {
    return `{${this.key}: ${this.value}}`;
  }
}

class LinkedListMap {
  constructor() {
    this.dummyHead = new Node(null, null);
    this.size = 0;
  }

  getSize() {
    return this.size;
  }

  isEmpty() {
    return this.size === 0;
  }

  getNode(key) {
    let cur = this.dummyHead.next;
    while(cur != null) {
      if (cur.key === key) {
        return cur;
      }
      cur = cur.next;
    }
    return null;
  }

  contains(key) {
    return this.getNode(key) !== null;
  }

  get(key) {
    const node = this.getNode(key);
    // 这里其实并不严谨，如果我们对应 key 的值就是 null，那这里其实是无法区分的
    // 但仔细想想虽然不严谨，但并不会出错，因为如果存在你查出来也是 null
    // 和不存在的值没区别
    if (node === null) {
      return null;
    } else {
      return node.value;
    }
  }

  add(key, value) {
    const node = this.getNode(key);
    if (node === null) {
      const newNode = new Node(key, value);
      newNode.next = this.dummyHead.next;
      this.dummyHead.next = newNode;
      // 记住这个，老忘
      this.size++;
    } else {
      node.value = value;
    }
  }

  set(key, value) {
    const node = this.getNode(key, value);
    if (node === null) {
      throw new Error("Key is not in map");
    } else {
      node.value = value;
    }
  }

  remove(key) {
    let prev = this.dummyHead;
    while(prev.next !== null) {
      if (key === prev.next.key) {
        let cur = prev.next;
        prev.next = cur.next;
        const value = cur.value;
        cur = null;
        this.size--;
        return value;
      }
      prev = prev.next;
    }

    return null;
  }

  // @TODO:
  toString() {
    return '11232131'
  }
}

export default LinkedListMap;

File: go/map/linked_list_map.go
package selfmap

import (
	"errors"
	"fmt"
)

type ListNode struct {
	key string
	value int
	next *ListNode
}

func (n *ListNode) String() string {
	return fmt.Sprintf("{%s: %d}", n.key, n.value)
}

type LinkedListMap struct {
	dummyHead *ListNode
	size int
}

func New() *LinkedListMap {
	return &LinkedListMap{&ListNode{}, 0}
}

func (m *LinkedListMap) GetSize() int {
	return m.size
}

func (m *LinkedListMap) IsEmpty() bool {
	return m.size == 0
}

func (m *LinkedListMap) getNode(key string) *ListNode {
	for cur := m.dummyHead.next; cur != nil; cur = cur.next {
		if cur.key == key {
			return cur
		}
	}
	return nil
}

func (m *LinkedListMap) Contains(key string) bool {
	return m.getNode(key) != nil
}

func (m *LinkedListMap) Get(key string) (value int, ok bool) {
	if node := m.getNode(key); node != nil {
		value = node.value
		ok = true
		return
	} else {
		ok = false
		return
	}
}

func (m *LinkedListMap) Add(key string, value int) {
	if node := m.getNode(key); node == nil {
		m.dummyHead.next = &ListNode{key: key, value: value, next: m.dummyHead.next,}
		m.size++
		return
	} else {
		node.value = value
	}
}

func (m *LinkedListMap) Set(key string, value int) error {
	if node := m.getNode(key); node == nil {
		return errors.New("Key doesn't exist")
	} else {
		node.value = value
		return nil
	}
}

func (m *LinkedListMap) Remove(key string) (value int, ok bool){
	prev := m.dummyHead;
	for ; prev.next != nil; prev = prev.next {
		if key == prev.next.key {
			cur := prev.next
			value = cur.value
			ok = true
			prev.next = cur.next
			cur = nil
			m.size--
			return
		}
	}
	return
}